market-generator/src/test/java/market/MarketSubscriptionManagerTest.java

```diff
@@ -0,0 +1,159 @@
+
+import java.util.HashSet;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@ExtendWith(MockitoExtension.class)
+class MarketSubscriptionManagerTest {
+
+    @InjectMocks
+    private MarketSubscriptionManager marketSubscriptionManager;
+
+    private final StreamObserver<TickResponse> subscriberOne = mock(StreamObserver.class);
+
+    private final StreamObserver<TickResponse> subscriberTwo = mock(StreamObserver.class);
+
+    private static final String GOOD_BANICA = "Banica";
+
+    private static final String GOOD_EGGS = "Eggs";
+
+    private static final MarketDataRequest MARKET_DATA_REQUEST_BANICA = MarketDataRequest.newBuilder().setGoodName(GOOD_BANICA).build();
+
+    private static final MarketDataRequest MARKET_DATA_REQUEST_INVALID = MarketDataRequest.newBuilder().build();
+
+
+    @Test
+    void subscribeShouldAddNewSubscriberToSubscriptions() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberOne);
+
+        int actual = marketSubscriptionManager.getSubscribers(MARKET_DATA_REQUEST_BANICA.getGoodName()).size();
+        int expected = 1;
+
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    void subscribeShouldNotAddNewSubscriberToSubscriptionsIfIllegalGoodName() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_INVALID, subscriberOne);
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberTwo);
+
+        int actual = marketSubscriptionManager.getSubscribers(MARKET_DATA_REQUEST_BANICA.getGoodName()).size();
+        int expected = 1;
+
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    void unsubscribeShouldNotRemoveSubscriberIfIllegalGoodName() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberOne);
+        marketSubscriptionManager.unsubscribe(MARKET_DATA_REQUEST_INVALID, subscriberOne);
+
+        int actual = marketSubscriptionManager.getSubscribers(MARKET_DATA_REQUEST_BANICA.getGoodName()).size();
+        int expected = 1;
+
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    void unsubscribeShouldUnsubscribeTheSubscriber() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberOne);
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberTwo);
+
+        marketSubscriptionManager.unsubscribe(MARKET_DATA_REQUEST_BANICA, subscriberTwo);
+
+        int actual = marketSubscriptionManager.getSubscribers(MARKET_DATA_REQUEST_BANICA.getGoodName()).size();
+        int expected = 1;
+
+        Assert.assertEquals(expected, actual);
+    }
+
+    @Test
+    public void unsubscribeShouldRemoveMapWhenOnlyOneSubscriberInside() {
+        marketSubscriptionManager.subscribe(MARKET_DATA_REQUEST_BANICA, subscriberOne);
```

Comment:
this test should be refactored as with the current implementation is hard to understand if you are testing subscribe or unsubscribe and should be re-structured as the comments mentioned above